### PR TITLE
Improve performance of distinct aggregations

### PR DESCRIPTION
**Pull request by Dith3r**
## Description

Improve performance of distinct aggregations by defining additional strategies based on source properties (for example, NDV).

The strategy to use for multiple distinct aggregations.
`SINGLE_STEP` Computes distinct aggregations in single-step without any pre-aggregations.
This strategy will perform poorly if the number of distinct grouping keys is small.
`MARK_DISTINCT` uses `MarkDistinct` for multiple distinct aggregations
or for mix of distinct and non-distinct aggregations.
`PRE_AGGREGATE` Computes distinct aggregations using a combination of aggregation 
and pre-aggregation steps.
`AUTOMATIC` chooses the strategy automatically.

Single-step strategy is preferred. However, for cases with limited concurrency due to
a small number of distinct grouping keys, it will choose an alternative strategy 
based on input data statistics.

| Strategy | Duration | Query |
| --- | --- | --- |
| MARK_DISTINCT | 5353ms | select count(ss_customer_sk), count(distinct ss_ticket_number) from hive.tpcds_sf1000_orc.store_sales;
| PRE_AGGREGATE | 2468ms | select count(ss_customer_sk), count(distinct ss_ticket_number) from hive.tpcds_sf1000_orc.store_sales;
| MARK_DISTINCT | 10109ms | select count(ss_quantity), count(distinct ss_item_sk), count(distinct ss_store_sk) from hive.tpcds_sf1000_orc.store_sales;
| PRE_AGGREGATE | 8253ms | select count(ss_quantity), count(distinct ss_item_sk), count(distinct ss_store_sk) from hive.tpcds_sf1000_orc.store_sales;

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```
# General
* Improve performance of distinct aggregations. ({issue}`21907`)
  Replaced configuration property `optimizer.mark-distinct-strategy` with `optimizer.distinct-aggregations-strategy`. 
  It's allowed values are changed as follows: `NONE -> SINGLE_STEP` and `ALWAYS -> MARK_DISTINCT`. 
  Replaced `optimizer.optimize-mixed-distinct-aggregations` configuration property 
  with a new `PRE_AGGREGATE` option in `optimizer.distinct-aggregations-strategy` .
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/SystemSessionProperties.java [hunk 1]
```diff
@@ -1370,9 +1370,9 @@ public static int getFilterAndProjectMinOutputPageRowCount(Session session)
         return session.getSystemProperty(FILTER_AND_PROJECT_MIN_OUTPUT_PAGE_ROW_COUNT, Integer.class);
     }
 
-    public static MarkDistinctStrategy markDistinctStrategy(Session session)
+    public static DistinctAggregationsStrategy distinctAggregationsStrategy(Session session)
     {
-        return session.getSystemProperty(MARK_DISTINCT_STRATEGY, MarkDistinctStrategy.class);
+        return session.getSystemProperty(DISTINCT_AGGREGATIONS_STRATEGY, DistinctAggregationsStrategy.class);
```
Comment (lukasz-stec) [inline core/trino-main/src/main/java/io/trino/SystemSessionProperties.java]: you should merge this commit with the commit that re-adds the MARK_DISTINCT_STRATEGY propertry

File: core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java [hunk 2]
```diff
@@ -683,6 +683,9 @@ public PlanOptimizers(
                                 new RemoveRedundantIdentityProjections(),
                                 new PushAggregationThroughOuterJoin(),
                                 new ReplaceRedundantJoinWithSource(), // Run this after PredicatePushDown optimizer as it inlines filter constants
+                                new DistinctAggregationToGroupBy(plannerContext), // Run this after aggregation pushdown so that multiple distinct aggregations can be pushed into a connector
+                                // It also is run before MultipleDistinctAggregationToMarkDistinct to take precedence f enabled
```
Comment (lukasz-stec) [inline core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java:687]: nit: typo should be _precedence if enabled_

**Review by lukasz-stec — APPROVED** [review]
lgtm % comments

File: core/trino-main/src/main/java/io/trino/sql/planner/OptimizerConfig.java [hunk 3]
```diff
@@ -399,18 +407,6 @@ public OptimizerConfig setIterativeOptimizerTimeout(Duration timeout)
         return this;
     }
 
-    public boolean isOptimizeMixedDistinctAggregations()
-    {
-        return optimizeMixedDistinctAggregations;
-    }
-
-    @Config("optimizer.optimize-mixed-distinct-aggregations")
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/OptimizerConfig.java:407]: Add to DefunctConfig ?

File: core/trino-main/src/main/java/io/trino/sql/planner/OptimizerConfig.java [hunk 4]
```diff
@@ -474,16 +474,16 @@ public OptimizerConfig setOptimizeMetadataQueries(boolean optimizeMetadataQuerie
     }
 
     @Nullable
-    public MarkDistinctStrategy getMarkDistinctStrategy()
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/OptimizerConfig.java:477]: It's confusing to see this config getting deleted here and then added back again in another commit. Can we avoid that ?

File: core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestMultipleDistinctAggregationToMarkDistinct.java [hunk 5]
```diff
@@ -171,14 +173,15 @@ public void testGlobalAggregation()
     @Test
     public void testAggregationNDV()
     {
-        PlanNodeId aggregationNodeId = new PlanNodeId("aggregationNodeId");
+        PlanNodeId aggregationId = new PlanNodeId("aggregationId");
```
Comment (raunaqmorarka) [inline core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestMultipleDistinctAggregationToMarkDistinct.java]: There should be no changes to existing plan tests in this commit right ?
Are these just just cosmetic changes ? If so, they shouldn't be in this commit

File: core/trino-main/src/main/java/io/trino/SystemSessionProperties.java [hunk 6]
```diff
@@ -119,8 +119,9 @@ public final class SystemSessionProperties
     public static final String DISTRIBUTED_SORT = "distributed_sort";
     public static final String USE_PARTIAL_TOPN = "use_partial_topn";
     public static final String USE_PARTIAL_DISTINCT_LIMIT = "use_partial_distinct_limit";
-    public static final String MAX_RECURSION_DEPTH = "max_recursion_depth";
     public static final String MARK_DISTINCT_STRATEGY = "mark_distinct_strategy";
+    public static final String MAX_RECURSION_DEPTH = "max_recursion_depth";
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/SystemSessionProperties.java:123]: Let's avoid moving this line, seems unrelated

File: core/trino-main/src/main/java/io/trino/SystemSessionProperties.java [hunk 7]
```diff
@@ -565,6 +561,12 @@ public SystemSessionProperties(
                         "Use partial Distinct Limit",
                         true,
                         true),
+                enumProperty(
+                        MARK_DISTINCT_STRATEGY,
+                        "",
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/SystemSessionProperties.java:566]: Please add DISTINCT_AGGREGATIONS_STRATEGY below the existing entry for MARK_DISTINCT_STRATEGY instead of moving MARK_DISTINCT_STRATEGY. The diff looks confusing right now.

File: core/trino-main/src/main/java/io/trino/SystemSessionProperties.java [hunk 8]
```diff
@@ -575,10 +577,10 @@ public SystemSessionProperties(
                         value -> validateIntegerValue(value, MAX_RECURSION_DEPTH, 1, false),
                         object -> object),
                 enumProperty(
-                        MARK_DISTINCT_STRATEGY,
+                        DISTINCT_AGGREGATIONS_STRATEGY,
                         "",
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/SystemSessionProperties.java:581]: Please add description message

**Review by raunaqmorarka — COMMENTED** [review]
lgtm % minor comments

**Comment by raunaqmorarka** [comment]
Please update the release notes section to specify which system session properties and configs were removed/defunct and what are the new alternatives.

File: core/trino-main/src/main/java/io/trino/SystemSessionProperties.java [hunk 9]
```diff
@@ -574,12 +582,6 @@ public SystemSessionProperties(
                         false,
                         value -> validateIntegerValue(value, MAX_RECURSION_DEPTH, 1, false),
                         object -> object),
-                enumProperty(
-                        MARK_DISTINCT_STRATEGY,
-                        "",
-                        MarkDistinctStrategy.class,
-                        optimizerConfig.getMarkDistinctStrategy(),
-                        false),
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/SystemSessionProperties.java:582]: Instead of deleting from here and added above, it should be modified here

**Review by raunaqmorarka — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java [hunk 10]
```diff
@@ -301,6 +303,7 @@ public PlanOptimizers(
                 costCalculatorWithEstimatedExchanges,
                 costComparator,
                 taskCountEstimator,
+                createDistinctAggregationStrategyChooser(taskCountEstimator),
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java:306]: Create directly this in the rules that need it -- just pass the TaskCountEstimator object. PlanOptimizers should not have to be aware about this implementation detail.

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DistinctAggregationStrategyChooser.java [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+        // can parallelize single-step, and single-step distinct is more efficient than alternatives
+        return true;
+    }
+
+    private int getMaxNumberOfConcurrentThreadsForAggregation(Session session)
+    {
+        return taskCountEstimator.estimateHashedTaskCount(session) * getTaskConcurrency(session);
+    }
+
+    private double getMinDistinctValueCountEstimate(AggregationNode aggregationNode, Rule.Context context)
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DistinctAggregationStrategyChooser.java:93]: Methods in this class only care about the session and stats provider, so pass them directly. In particular, this class should not have to be aware about the rule evaluation framework (i.e., Rule.Context)

File: core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java [hunk 12]
```diff
@@ -46,6 +46,7 @@
 import io.trino.sql.planner.iterative.rule.DetermineJoinDistributionType;
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java:46]: commit message is too long. Do we need rename in the first place?

File: core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java [hunk 13]
```diff
@@ -682,6 +682,9 @@ public PlanOptimizers(
                                 new RemoveRedundantIdentityProjections(),
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java:682]: Update commit message and description
